### PR TITLE
AB#1052 Add an update log

### PR DESCRIPTION
**Pull request by daniel-weisse**
### Proposed changes
Add a log of updates to a running coordinator instance.
The log can be requested using GET requests on the `/update` endpoint.

Also comes with an integration for the CLI, viewable using the `--update-log` or `-u` flag for the `marblerun manifest get` command:
```bash
$ marblerun manifest get $MARBLERUN --update-log
Successfully verified coordinator, now requesting manifest signature
Manifest signature: 152493c4a85845480a04b95f79dd447a4573862e0d2c102c71b91b8b3cbcade5
Update History:
2021/06/25 13:11:28UTC original manifest set
2021/06/25 13:12:41UTC [admin] set secret [symmetric_key_unset]
2021/06/25 13:12:41UTC [admin] set secret [cert_unset]
2021/06/25 13:14:24UTC [admin] updated package [world] to SecurityVersion 2
2021/06/25 13:41:12UTC [admin] updated package [world] to SecurityVersion 4
```

### CODE REVIEW

**Comment by thomasten** [comment]
The log should be machine readable such that it is easy to create a "virtual manifest" based on the initial manifest and the log. We may use zaplogger's strucutured logging capabilites for this.

**Comment by daniel-weisse** [comment]
The update log is now written using zap.
Sample output from `marblerun manifest get $MARBLERUN --update-log`:
```
Successfully verified coordinator, now requesting manifest signature
Manifest signature: 152493c4a85845480a04b95f79dd447a4573862e0d2c102c71b91b8b3cbcade5
{"time":"2021-06-28T13:43:30.239Z","update":"original manifest set"}
{"time":"2021-06-28T13:43:48.971Z","update":"SecurityVersion increased","user":"admin","package":"world","new version":4}
{"time":"2021-06-28T13:44:29.907Z","update":"secret set","user":"admin","secret":"cert_unset","type":"cert-ed25519"}
{"time":"2021-06-28T13:44:29.907Z","update":"secret set","user":"admin","secret":"symmetric_key_unset","type":"symmetric-key"}
```

Each entry contains the current time in ISO 8601 format, an update message, and additional information based on the updated resource.
For secrets that is:
* the user setting the secret
* the secrets name
* the secrets type

For packages:
* the user performing the update
* the name of the package
* the new security version number

Other metadata, such as log-level and caller, is omitted in the log, since it is not relevant for its purpose.
Please let me know if this format works for the update log or if it should be changed.
For reference: the standard logger of the core uses an unmodified `zap.NewProduction` logger.
This includes a timestamp in seconds, the log level and caller information (module and line of code).

Since the update log is mostly supposed to be read by humans i prefer the timestamp in a human readable format.

File: coordinator/core/clientapi.go [hunk 1]
```diff
@@ -296,6 +311,15 @@ func (c *Core) UpdateManifest(ctx context.Context, rawUpdateManifest []byte, upd
 		return err
 	}
 
+	oldLog, err := c.data.getUpdateLog()
+	if err != nil {
+		return err
+	}
+	c.updatelog.reset(oldLog)
+	for pkgName, pkg := range updateManifest.Packages {
+		c.updatelog.logger.Info("SecurityVersion increased", zap.String("user", updater.Name()), zap.String("package", pkgName), zap.Uint("new version", *pkg.SecurityVersion))
```
Comment (thomasten) [inline coordinator/core/clientapi.go:320]: That seems cumbersome. I think the updateLog code should be in its own package and be usable like this:
```go
updateLogger := updateLog.New()
updateLogger.Info(...) // should work by embedding zap.Logger into the updateLogger struct
txdata.appendUpdateLog(updateLogger.String())
```

**Review by thomasten — CHANGES_REQUESTED** [review]

File: cli/cmd/manifestGet.go [hunk 2]
```diff
@@ -29,48 +31,77 @@ func newManifestGet() *cobra.Command {
...
 			fmt.Printf("Manifest signature: %s\n", string(response))
+			if updates {
+				fmt.Println(string(updateLog))
+			}
 			return nil
 		},
 		SilenceUsage: true,
 	}
 	cmd.Flags().StringVarP(&output, "output", "o", "", "Save singature to file instead of printing to stdout")
+	cmd.Flags().BoolVarP(&updates, "update-log", "u", false, "Print a log of updates to the coordinator")
```
[the remark this answers is not in the document]
Comment (m1ghtym0) [reply cli/cmd/manifestGet.go:72]: Yes, we should not mix the manifest get and log into one command.
Could be `marblerun manifest log` for example.

[the remark this answers is not in the document]
Comment (daniel-weisse) [reply cli/cmd/manifestGet.go:72]: sounds good

**Review by thomasten — APPROVED** [review]

**Pull request merged**